### PR TITLE
Add clique consensus

### PR DESCRIPTION
**Pull request by lupin012**
This PR introduces the Clique consensus engine to correctly set the reward beneficiary for `trace/debug` APIs in Goerli network

### CODE REVIEW

File: core/silkworm/common/base.hpp [hunk 1]
```diff
@@ -75,6 +75,8 @@ inline constexpr size_t kAddressLength{20};
 
 inline constexpr size_t kHashLength{32};
 
+inline constexpr size_t kEntraSealSize{65};
```
Comment (canepat) [inline core/silkworm/common/base.hpp:78]: There is a typo here, it should read `kExtraSealSize`

File: core/silkworm/consensus/clique/engine.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    return ecrecover(header);
+}
+
+evmc::address
+CliqueEngine::ecrecover(const BlockHeader& header) {
+    evmc::address beneficiary = evmc::address{};
+
+    evmc::bytes32 seal_hash = header.hash(false, true);
+    Bytes signature = header.extra_data.substr(header.extra_data.length() - kEntraSealSize, kEntraSealSize - 1);
+    bool odd_y_parity = header.extra_data[header.extra_data.length() - 1] == 1;
```
Comment (canepat) [inline core/silkworm/consensus/clique/engine.cpp:35]: I suggest changing this to:

`bool odd_y_parity = header.extra_data[header.extra_data.length() - 1] != 0;`

File: core/silkworm/consensus/clique/engine.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include "engine.hpp"
+
+#include <silkpre/ecdsa.h>
+
+namespace silkworm::consensus {
+
+ValidationResult CliqueEngine::validate_seal(const BlockHeader&) { return ValidationResult::kOk; }
+
+evmc::address CliqueEngine::get_beneficiary(const BlockHeader& header) {
+    return ecrecover(header);
```
Comment (canepat) [inline core/silkworm/consensus/clique/engine.cpp:26]: I suggest adding a check here on `extra_data` size and fall back to parent `get_beneficiary` implementation if not satisfied

File: core/silkworm/consensus/engine_test.cpp [hunk 4]
```diff
@@ -61,6 +63,12 @@ TEST_CASE("Consensus Engine Seal") {
     CHECK(consensus_engine->validate_seal(fake_header) == ValidationResult::kOk);
 }
 
+TEST_CASE("Consensus Engine Clique") {
+    BlockHeader fake_header{};
+    auto consensus_engine = engine_factory(kGoerliConfig);  // Clique consensus engine
+    CHECK(consensus_engine->validate_seal(fake_header) == ValidationResult::kOk);
+}
+
```
Comment (canepat) [inline core/silkworm/consensus/engine_test.cpp:71]: I would move this test case to a `engine_test.cpp` module dedicated to Clique consensus

File: core/silkworm/types/block.hpp [hunk 5]
```diff
@@ -55,7 +55,7 @@ struct BlockHeader {
 
     std::optional<intx::uint256> base_fee_per_gas{std::nullopt};  // EIP-1559
 
-    [[nodiscard]] evmc::bytes32 hash(bool for_sealing = false) const;
+    [[nodiscard]] evmc::bytes32 hash(bool for_sealing = false, bool eip225 = false) const;
```
Comment (canepat) [inline core/silkworm/types/block.hpp:58]: I suggest renaming `eip225` to `exclude_extra_data_sig`

File: core/silkworm/types/block.hpp [hunk 6]
```diff
@@ -91,7 +91,7 @@ namespace rlp {
     size_t length(const Block&);
 
     void encode(Bytes& to, const BlockBody&);
-    void encode(Bytes& to, const BlockHeader&, bool for_sealing = false);
+    void encode(Bytes& to, const BlockHeader&, bool for_sealing = false, bool eip225 = false);
```
Comment (canepat) [inline core/silkworm/types/block.hpp:94]: I suggest renaming `eip225` to `exclude_extra_data_sig`

File: core/silkworm/types/block.cpp [hunk 7]
```diff
@@ -44,7 +44,7 @@ void Block::recover_senders() {
 
 namespace rlp {
 
-    static Header rlp_header(const BlockHeader& header, bool for_sealing = false) {
+    static Header rlp_header(const BlockHeader& header, bool for_sealing = false, bool eip225 = false) {
```
Comment (canepat) [inline core/silkworm/types/block.cpp:47]: I suggest renaming `eip225` to `exclude_extra_data_sig`

**Review by canepat — CHANGES_REQUESTED** [review]

File: core/silkworm/consensus/engine_test.cpp (hunk 4 above)
Comment (lupin012) [reply core/silkworm/consensus/engine_test.cpp]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**